apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/CallTree.java

```diff
@@ -436,7 +436,16 @@ private void fillStackTrace(List<StackFrame> stackTrace) {
         }
     }
 
-    public void recycle(ObjectPool<CallTree> pool) {
+    /**
+     * Recycles this subtree to the provided pool recursively.
+     * Note that this method ends by recycling {@code this} node (i.e. - this subtree root), which means that
+     * <b>the caller of this method should make sure that no reference to this object is held anywhere</b>.
+     * <p>ALSO NOTE: MAKE SURE NOT TO CALL THIS METHOD FOR {@link CallTree.Root} INSTANCES.</p>
+     *
+     * @param pool the pool to which all subtree nodes are to be recycled
+     */
+    public final void recycle(ObjectPool<CallTree> pool) {
+        assert !(this instanceof Root);
```

Comment:
why not throw directly an `IllegalArgumentException` when called on a root node instead of an assertion that might be disabled ?

Reply:
I thought of avoiding the `instanceof` check overhead in production. Normally I think it is not something that would be specific for a unique corner case, our tests should find it if we try to recycle improperly